cbpfc.go

```diff
@@ -467,17 +477,25 @@ func addIndirectPacketGuards(blocks []*block) {
 }
 
 type packetGuardOpts struct {
-	// requiredGuard returns:
-	// - the packetGuard needed by insns
-	// - the number of instructions in insns covered by the guard.
-	//   The guard is assumed to be invalidated for the remaining / uncovered insns (eg RegX was clobbered for indirect guards).
-	//   requiredGuard will be called until all instructions are covered.
-	requiredGuard func(insns []instruction) (int, packetGuard)
+	// requiredGuard returns the packetGuard needed by insns, and what insns it is valid for.
+	requiredGuard func(insns []instruction) requiredGuard
 
 	// createInsn creates an instruction that checks the packet length against the guard
 	createInsn func(guard packetGuard) bpf.Instruction
 }
 
+type requiredGuard struct {
+	guard packetGuard
+
+	// The guard covers all the requested instructions,
+	// and is still valid afterwards.
+	alwaysValid bool
+
+	// The guard covers n instructions,
+	// and isn't valid for the subsequent n+1: instructions (eg RegX was clobbered for indirect guards).
```

Comment:
For my understanding: does the clobber always happen on the last insn?

Reply:
Yup. That invalidates the guard, and we call `requiredGuard` again for the remaining instructions to see if they need one.